strutil/env.go

```diff
@@ -0,0 +1,253 @@
+func (env *Environment) Transform(tr func(key, value string) (string, string)) {
+	newEntries := make(map[string]string, len(env.entries))
+	keys := make([]string, 0, len(env.entries))
+	for key := range env.entries {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		newKey, newValue := tr(key, env.entries[key])
+		if newKey != "" {
+			newEntries[newKey] = newValue
+		}
+	}
+	env.entries = newEntries
+}
+
+// Get returns the value of a given environment variable.
+func (env *Environment) Get(key string) string {
+	return env.entries[key]
+}
+
+// Contains returns true if given environment variable is set.
+func (env *Environment) Contains(key string) bool {
+	_, ok := env.entries[key]
+	return ok
+}
+
+// Del removes the given environment variable.
+func (env *Environment) Del(key string) {
+	delete(env.entries, key)
+}
+
+// Set adds or replaces the given environment variable.
+func (env *Environment) Set(key, value string) {
+	if env.entries == nil {
+		env.entries = make(map[string]string)
+	}
+	env.entries[key] = value
+}
+
+// RawEnvironment returns the raw equivalent of the given environment.
+// The returned environment is sorted lexicographically by variable name.
+func (env *Environment) RawEnvironment() RawEnvironment {
+	raw := make([]string, 0, len(env.entries))
+	keys := make([]string, 0, len(env.entries))
+	for key := range env.entries {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		raw = append(raw, fmt.Sprintf("%s=%s", key, env.entries[key]))
+	}
+	return raw
+}
+
+// EnvironmentDelta describes alterations to an environment.
+//
+// Differential environment can refer to existing entries by using shell-like
+// syntax $KEY or ${KEY}. Entries inside an environment delta are ordered.
+type EnvironmentDelta struct {
+	OrderedMap
+}
+
+// NewEnvironmentDelta returns a new environment delta comprised of given pairs.
+func NewEnvironmentDelta(pairs ...string) *EnvironmentDelta {
+	return &EnvironmentDelta{OrderedMap: *NewOrderedMap(pairs...)}
+}
+
+// Copy returns a copy of the environment delta.
+func (delta *EnvironmentDelta) Copy() *EnvironmentDelta {
+	return &EnvironmentDelta{OrderedMap: *delta.OrderedMap.Copy()}
+}
+
+// Merge combines two environment deltas.
+//
+// Clashing environment variables are overwritten and the value from the
+// "other" delta prevails.
+func (delta *EnvironmentDelta) Merge(other *EnvironmentDelta) {
+	for _, key := range other.Keys() {
+		value := other.Get(key)
+		delta.Set(key, value)
+	}
+}
+
+// ApplyDelta applies a delta to the environment.
+//
+// Environment is modified in place. Delta acts like an environment that can
+// contain additional key=value pairs, where values can refer to any key set in
+// the environment or eventually computed by applying the delta.
+//
+// The return value is the ordered list of variables that were referenced by
+// the delta but were never defined. They are expanded to an empty string.
```

Comment:
The doc comment is hard to read (probably because it doesn't mention what refer here means), this looks more powerful than SubstituteEnv, is there are a deep reason for that?

Reply:
My initial implementation took varadic list of deltas to apply and I wanted it to be able to successfully resolve the scenario where PATH is defined in the environment and then expanded by per-snap and per-app definitions.

On second though I think that the way delta Merge works this property is not retained anymore. I think the stable-state loop could be discarded without observable effects.

This does beg the question of what ought to happen in this scenario:
```
name: test-snapd-env-expansion
version: 1
environment:
  PATH: $PATH:$SNAP/global
apps:
  app:
    environment:
      PATH: $PATH:$SNAP/local
```

Currently, as implemented here PATH will be only `$PATH/$SNAP/local`, is that what we want?

Reply:
A version with weaker ApplyDelta for consideration: https://paste.ubuntu.com/p/9zYcXtXJS7/

Reply:
the question is really whether app environment overrides should be resolved using only the os environment and base environment, or also the test snap environment overrides

Reply:
Indeed. I believe we have under-specified things so it was de-facto defined by the first implementation. I hope it's not too late to define and document something sensible.

Reply:
I think substituting down chain from os env and base env to global env to app/hook env is what we want, but then we really don't want current behavior here that solves variables with later values

Reply:
We now have the chaining semantics. I will still tweak Apply to have simpler logic and take many ExpandableEnvs as arguments.

Reply:
Done